remote-config/src/fetch/fetcher.rs

```diff
@@ -0,0 +1,775 @@
+        contents: Vec<u8>,
+    ) -> anyhow::Result<()>;
+}
+
+/// Fundamental configuration of the RC client, which always must be set.
+#[derive(Clone, Hash, Eq, PartialEq)]
+pub struct ConfigInvariants {
+    pub language: String,
+    pub tracer_version: String,
+    pub endpoint: Endpoint,
+    pub products: Vec<RemoteConfigProduct>,
+    pub capabilities: Vec<RemoteConfigCapabilities>,
+}
+
+struct StoredTargetFile<S> {
+    hash: String,
+    handle: Arc<S>,
+    state: ConfigState,
+    meta: TargetFileMeta,
+    expiring: bool,
+}
+
+pub struct ConfigFetcherState<S> {
+    target_files_by_path: Mutex<HashMap<String, StoredTargetFile<S>>>,
+    pub invariants: ConfigInvariants,
+    endpoint: Endpoint,
+    pub expire_unused_files: bool,
+}
+
+pub struct ConfigFetcherFilesLock<'a, S> {
+    inner: MutexGuard<'a, HashMap<String, StoredTargetFile<S>>>,
+}
+
+impl<'a, S> ConfigFetcherFilesLock<'a, S> {
+    /// Actually remove the file from the known files.
+    /// It may only be expired if already marked as expiring.
+    pub fn expire_file(&mut self, path: &RemoteConfigPath) {
+        if let Entry::Occupied(entry) = self.inner.entry(path.to_string()) {
+            if entry.get().expiring {
+                entry.remove();
+            }
+        }
+    }
+
+    /// Stop advertising the file as known. It's the predecessor to expire_file().
+    pub fn mark_expiring(&mut self, path: &RemoteConfigPath) {
+        if let Some(target_file) = self.inner.get_mut(&path.to_string()) {
+            target_file.expiring = true;
+        }
+    }
+}
+
+impl<S> ConfigFetcherState<S> {
+    pub fn new(invariants: ConfigInvariants) -> Self {
+        ConfigFetcherState {
+            target_files_by_path: Default::default(),
+            endpoint: get_product_endpoint(PROD_INTAKE_SUBDOMAIN, &invariants.endpoint),
+            invariants,
+            expire_unused_files: true,
+        }
+    }
+
+    /// To remove unused remote files manually. Must not be called when auto expiration is active.
+    /// Note: careful attention must be paid when using this API in order to not deadlock:
+    /// - This files_lock() must always be called prior to locking any data structure locked within
+    ///   FileStorage::store().
+    /// - Also, files_lock() must not be called from within FileStorage::store().
+    pub fn files_lock(&self) -> ConfigFetcherFilesLock<S> {
+        assert!(!self.expire_unused_files);
+        ConfigFetcherFilesLock {
+            inner: self.target_files_by_path.lock().unwrap(),
+        }
+    }
+}
+
+pub struct ConfigFetcher<S: FileStorage> {
+    pub file_storage: S,
+    state: Arc<ConfigFetcherState<S::StoredFile>>,
+    /// Timeout after which to report failure, in milliseconds.
+    pub timeout: AtomicU32,
+    /// Collected interval. May be zero if not provided by the remote config server or fetched yet.
+    /// Given in nanoseconds.
+    pub interval: AtomicU64,
+}
+
+#[derive(Default)]
+pub struct OpaqueState {
```

Comment:
This state is opaque and not exposed outside of RC backends remove it.

Reply:
The implementing document says about the opaque_backend_state:
> MUST contain the opaque_backend_state field extracted from targets. 
> This is used to make the backend stateless by saving some opaque lightweight data directly in the agent and tracers.
> For tracer clients this is mainly used for accurate tracking of where configurations are sent.

I.e. that I'll have to extract and submit it unmodified with each request?